File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/12. 7576.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const [M, N] = inputArr[0].trim().split(' ').map(Number)
const boxMatrix = inputArr.slice(1).map(row => row.trim().split(' ').map(Number))

// 며칠만에 익었는지 저장할 배열
const isVisit = Array.from({length: N}, () => Array.from({length: M}).fill(-1))

// BFS 용 큐 역할을 할 배열
const visitQueue = Array.from({length: 1000001}).fill([])
let nowIdx = 0
let insertIdx = 0

// 이미 익어있는 토마토들에 대한 연산(0일만에 익음 + 방문큐에 저장)
for (let i = 0; i < N; i++) {
  for (let j = 0; j < M; j++) {
    // 익은 토마토가 있거나 빈 공간은 0일만에 익었다고 간주한다.
    if (boxMatrix[i][j] === 1 || boxMatrix[i][j] === -1) {
      isVisit[i][j] = 0

      // 익은 토마토가 있는 공간만 방문해야 한다.
      if (boxMatrix[i][j] === 1) {
        visitQueue[insertIdx++] = [i, j]
      }
    }
  }
}

// 최대값 저장할 변수
let result = 0

while (nowIdx < insertIdx) {
  const [nowRow, nowCol] = visitQueue[nowIdx++]
  const nowDate = isVisit[nowRow][nowCol]

  result = Math.max(result, nowDate)

  if (nowRow > 0 && boxMatrix[nowRow - 1][nowCol] === 0 && isVisit[nowRow - 1][nowCol] === -1) {
    isVisit[nowRow - 1][nowCol] = nowDate + 1
    visitQueue[insertIdx++] = [nowRow - 1, nowCol]
  }
  if (nowRow < N - 1 && boxMatrix[nowRow + 1][nowCol] === 0 && isVisit[nowRow + 1][nowCol] === -1) {
    isVisit[nowRow + 1][nowCol] = nowDate + 1
    visitQueue[insertIdx++] = [nowRow + 1, nowCol]
  }
  if (nowCol > 0 && boxMatrix[nowRow][nowCol - 1] === 0 && isVisit[nowRow][nowCol - 1] === -1) {
    isVisit[nowRow][nowCol - 1] = nowDate + 1
    visitQueue[insertIdx++] = [nowRow, nowCol - 1]
  }
  if (nowCol < M - 1 && boxMatrix[nowRow][nowCol + 1] === 0 && isVisit[nowRow][nowCol + 1] === -1) {
    isVisit[nowRow][nowCol + 1] = nowDate + 1
    visitQueue[insertIdx++] = [nowRow, nowCol + 1]
  }
}

// 방문안한 칸의 갯수 구하는 부분
let howManyNull = isVisit.map(rowArr => rowArr.filter(val => val === -1).length).reduce((prev, val) => (prev += val), 0)

// 결과 출력
if (howManyNull > 0) {
  process.stdout.write('-1')
} // ::
else {
  process.stdout.write(`${result}`)
}

File: 1. 단계별로_풀어보기/20~29/27. 그래프와 순회/12. 7576.py
from collections import deque
import sys

input = sys.stdin.readline
output = sys.stdout.write

M, N = map(int, input().split())
boxMatrix = list(list(map(int, input().split())) for i in range(N))

isVisit = list(list(-1 for i in range(M)) for j in range(N))

visitQueue = deque()

for row in range(N):
  for col in range(M):
    # 익은 토마토는 0일만에 익었음을 표시 + 방문큐에 저장
    if boxMatrix[row][col] == 1:
      isVisit[row][col] = 0
      visitQueue.append([row, col])
    # 빈 공간은 0일만에 익었다고 간주
    elif boxMatrix[row][col] == -1:
      isVisit[row][col] = 0

result = 0

# BFS 연산
while len(visitQueue) > 0:
  nowRow, nowCol = visitQueue.popleft()
  nowDate = isVisit[nowRow][nowCol]
  result = max(result, nowDate)

  if nowRow > 0 and boxMatrix[nowRow - 1][nowCol] == 0 and isVisit[nowRow - 1][nowCol] == -1:
    isVisit[nowRow - 1][nowCol] = nowDate + 1
    visitQueue.append([nowRow - 1, nowCol])

  if nowRow < N - 1 and boxMatrix[nowRow + 1][nowCol] == 0 and isVisit[nowRow + 1][nowCol] == -1:
    isVisit[nowRow + 1][nowCol] = nowDate + 1
    visitQueue.append([nowRow + 1, nowCol])

  if nowCol > 0 and boxMatrix[nowRow][nowCol - 1] == 0 and isVisit[nowRow][nowCol - 1] == -1:
    isVisit[nowRow][nowCol - 1] = nowDate + 1
    visitQueue.append([nowRow, nowCol - 1])
    
  if nowCol < M - 1 and boxMatrix[nowRow][nowCol + 1] == 0 and isVisit[nowRow][nowCol + 1] == -1:
    isVisit[nowRow][nowCol + 1] = nowDate + 1
    visitQueue.append([nowRow, nowCol + 1])


howManyNull = sum(len(list(1 for col in range(M) if isVisit[row][col] == -1)) for row in range(N))

if howManyNull > 0:
  output(f"-1")
else:
  output(f"{result}")